crates/cubecl-ir/src/type_hash.rs

```diff
@@ -0,0 +1,202 @@
+use std::borrow::ToOwned;
+use std::hash::Hasher;
+
+/// A hash of a type's structure
+pub trait TypeHash {
+    /// Generate a hash of the type structure.
+    #[allow(unused)]
+    fn type_hash() -> u64 {
+        let mut hasher = fnv::FnvHasher::default();
+        Self::write_hash(&mut hasher);
+        hasher.finish()
+    }
```

Comment:
What's the difference between a type hash and the `core::any::TypeId` ? Would hashing `core::any::type_name` would have the same effect?

Reply:
`TypeId` isn't stable across compilations, the point of a type hash is to version the structure so you can ensure a cache with binary serialization is valid for the current version of the IR. Otherwise you might get runtime errors or incorrect data.

Reply:
I added more documentation to make this clearer.